core/src/dird/dird_conf.cc

```diff
@@ -3080,12 +3080,16 @@ static void StoreRunscript(LEX* lc, ResourceItem* item, int index, int pass)
  * %C = Cloning (yes/no)
  * %D = Director name
  * %V = Volume name(s) (Destination)
+ * %m = Prev Backup JobId
+ * %M = New Backup JobId
  */
 extern "C" char* job_code_callback_director(JobControlRecord* jcr,
                                             const char* param)
 {
   static char yes[] = "yes";
   static char no[] = "no";
+  
+  static char str[50];
```

Comment:
Is using a static array to store and return the result okay? I copied that from another similar callback, so I'm not sure.

Reply:
I don't really like it, because it won't be thread-safe. i.e. if the threads for multiple jobs enter `job_code_callback()` around the same time, there's a possible data-race on `str` which may lead to the wrong value being returned.

However, there's no great way to work around this inside the function right now. I guess the best way to proceed would be to remove the requirement for `extern "C"` and return a `std::string` instead of a `char*`.
I'll see if I can make that change easily.

Reply:
Alright, thanks for your feedback. I already have a branch where I've refactored this callback to return an std::optional<std::string>, should I merge that into this PR or do you want that somewhere else?

Reply:
I've only refactored it up to util.cc#edit_job_codes though (https://github.com/SamuelBoerlin/bareos/commit/0b7498a18f66778f1612d003ad485285a0ee3164#diff-edeceb51c3de23d07b2d5d9e679d111e1238b74cc870a492143e5283e4386ee9R937-R943), after which point it gets more complicated.

Reply:
That looks promising! You can just add it to this PR as it is the refactoring required to implement the feature you're working on.

You could
- drop the `extern "C"` everywhere, having C-linkage if you need a `std::string` probably won't help much
- return the plain `char*`/`const char*`, compiler will take care of conversion on return

Your solution for `edit_job_codes()` is probably as good as it gets without refactoring the whole thing. However, for a plain `char*` `snprintf()` is better suited than `Bsnprintf()`, as the latter expects a resizable `POOLMEM*`.